lib/coarse_op.cuh

```diff
@@ -624,58 +172,287 @@ namespace quda {
       default:
 	errorQuda("Undefined compute type %d", type);
       }
+
       return threads;
     }
 
     bool tuneGridDim() const { return false; } // don't tune the grid dimension
     bool tuneAuxDim() const { return type != COMPUTE_VUV ? false : true; }
 
     unsigned int sharedBytesPerBlock(const TuneParam &param) const {
-      if (arg.shared_atomic && type == COMPUTE_VUV)
+      if (/*arg.shared_atomic && */type == COMPUTE_VUV)
         return 4*sizeof(storeType)*arg.max_color_height_per_block*arg.max_color_width_per_block*4*coarseSpin*coarseSpin;
-      return TunableVectorYZ::sharedBytesPerBlock(param);
-    }
-
-    unsigned int sharedBytesPerThread() const {
-      return TunableVectorYZ::sharedBytesPerThread();
+      return TunableKernel3D::sharedBytesPerBlock(param);
     }
 
   public:
-    CalculateY(Arg &arg, const ColorSpinorField &meta, GaugeField &Y, GaugeField &X, GaugeField &Y_atomic,
-               GaugeField &X_atomic, bool use_mma) :
-      TunableVectorYZ(2, 1),
+    CalculateY(Arg &arg, const ColorSpinorField &V, GaugeField &Y, GaugeField &X, GaugeField &Y_atomic,
+               GaugeField &X_atomic) :
+      TunableKernel3D(V, 2, 1),
       arg(arg),
-      type(COMPUTE_INVALID),
-      meta(meta),
+      V(V),
       Y(Y),
       X(X),
       Y_atomic(Y_atomic),
       X_atomic(X_atomic),
       dim(0),
       dir(QUDA_BACKWARDS),
-      use_mma(use_mma)
+      type(COMPUTE_INVALID),
+      compute_max(false)
     {
-      if (meta.Location() == QUDA_CUDA_FIELD_LOCATION) {
-#ifdef JITIFY
-        create_jitify_program("kernels/coarse_op_kernel.cuh");
+      strcat(aux, comm_dim_partitioned_string());
+    }
+
+    /**
+       @brief Launcher for CPU instantiations of coarse-link construction
+    */
+    template <QudaFieldLocation location_> std::enable_if_t<location_ == QUDA_CPU_FIELD_LOCATION>
+    Launch(Arg &arg, TuneParam &tp, ComputeType type, const qudaStream_t &stream)
+    {
+      if (compute_max) {
+        memset(arg.max_h, 0, sizeof(typename Arg::Float));
+        arg.max = arg.max_h;
+      }
+
+      if (type == COMPUTE_UV) {
+        if (compute_max) launch_host<compute_uv>(tp, stream, ArgMax<Arg>(arg));
+        else launch_host<compute_uv>(tp, stream, arg);
+      } else if (type == COMPUTE_AV) {
+        if (from_coarse) errorQuda("compute_av should only be called from the fine grid");
+
+#if defined(GPU_CLOVER_DIRAC) && defined(WILSONCOARSE)
+        if (compute_max) launch_host<compute_av>(tp, stream, ArgMax<Arg>(arg));
+        else launch_host<compute_av>(tp, stream, ArgMax<Arg>(arg));
```

Comment:
No `ArgMax`

Reply:
What does comment mean?

Reply:
This line should simply be `launch_host<compute_av>(tp, stream, arg);`, i.e., no `ArgMax`

Reply:
ah gotcha, good catch 😉

Reply:
9ffa88b04440e3406261ba56f8717cc1437ef7d4